polynote-kernel/src/main/scala/polynote/kernel/PolyKernel.scala

```diff
@@ -29,78 +31,109 @@ import scala.tools.nsc.interactive.Global
+  protected lazy val symbolTable = new RuntimeSymbolTable(kernelContext, statusUpdates)
 
   /**
     * The task queue, which tracks currently running and queued kernel tasks (i.e. cells to run)
     */
   protected lazy val taskQueue: TaskQueue = TaskQueue(statusUpdates).unsafeRunSync()
 
   // TODO: duplicates logic from runCell
-  private def runPredef(kernel: LanguageKernel[IO], language: String): IO[Unit] =
-    kernel.predefCode.fold(kernel.init()) {
-      code => taskQueue.runTaskIO(s"Predef $language", s"Predef ($language)")(_ => kernel.runCode("Predef", Nil, Nil, code)).flatMap {
-        results => results.evalMap {
-          case WrapSymbol(symbol) =>
-            symbolTable.publishAll(symbolTable.RuntimeValue.fromSymbolDecl(symbol).toList)
-          case _ => IO.unit
-        }.compile.drain
-      }
-    }
+  private def runPredef(interp: LanguageInterpreter[IO], language: String): IO[Stream[IO, Result]] =
+    interp.init() *> {
+      interp.predefCode match {
+        case Some(code) => taskQueue.runTaskIO(s"Predef $language", s"Predef ($language)")(_ => interp.runCode("Predef", Nil, Nil, code)).map {
+          results => results.collect {
+            case v: ResultValue => v
+          }.evalTap {
+            v => symbolTable.publishAll(List(symbolTable.RuntimeValue.fromResultValue(v, interp)))
+          }
+        }
 
-  protected def getKernel(language: String): IO[LanguageKernel[IO]] = Option(kernels.get(language)).map(IO.pure).getOrElse {
-    launchingKernel.acquire.bracket { _ =>
-      Option(kernels.get(language)).map(IO.pure).getOrElse {
-        for {
-          factory <- IO.fromEither(Either.fromOption(subKernels.get(language), new RuntimeException(s"No kernel for language $language")))
-          kernel  <- taskQueue.runTask(s"Kernel$$$language", s"Starting $language kernel")(_ => factory.apply(dependencies.getOrElse(language, Nil), symbolTable))
-          _        = kernels.put(language, kernel)
-          _       <- runPredef(kernel, language)
-        } yield kernel
+        case None => IO.pure(Stream.empty)
       }
-    } {
-      _ => launchingKernel.release
     }
+
+  protected def getInterpreter(language: String): Option[LanguageInterpreter[IO]] = Option(interpreters.get(language))
+
+  protected def getOrLaunchInterpreter(language: String): IO[(LanguageInterpreter[IO], Stream[IO, Result])] =
+    Option(interpreters.get(language))
+      .map(interp => IO.pure(interp -> Stream.empty))
+      .getOrElse {
+        launchingInterpreter.acquire.bracket { _ =>
+          Option(interpreters.get(language)).map(interp => IO.pure(interp -> Stream.empty)).getOrElse {
+            for {
+              factory <- IO.fromEither(Either.fromOption(subKernels.get(language), new RuntimeException(s"No interpreter for language $language")))
+              interp  <- taskQueue.runTask(s"Kernel$$$language", s"Starting $language interpreter")(_ => factory.apply(dependencies.getOrElse(language, Nil), symbolTable))
```

Comment:
>` s"Kernel$$$language"`

We should check to see whether this is a magic string like you suspected. Maybe a TODO so we don't forget?

Reply:
(and if it's not magic we should change to Interpreter)

Reply:
I couldn't find any evidence of magic, so I changed it.